cqrs-codegen/src/event.rs

```diff
@@ -0,0 +1,167 @@
+use proc_macro2;
+
+use quote::quote;
+
+use syn::{
+    self,
+    parse::{Error, Result},
+    spanned::Spanned as _,
+};
+
+use synstructure;
+
+use crate::utility;
+
+// ...
+
+pub fn event(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    match input.data {
+        syn::Data::Struct(_) => derive_struct(input),
+        syn::Data::Enum(_) => derive_enum(input),
+        syn::Data::Union(data) => Err(Error::new(data.union_token.span(), "Event can only be derived for structs and enums")),
+    }
+}
+
+
+fn derive_struct(input: syn::DeriveInput) -> Result<proc_macro2::TokenStream> {
+    let syn::DeriveInput {
+        attrs,
+        ident,
+        generics,
+        ..
+    } = input;
+
+    let meta = utility::get_nested_meta(attrs, "event")?
+        .ok_or_else(|| Error::new(proc_macro2::Span::call_site(), "Expected to find #[event(...)] attribute"))?;
+
+    let event_type = parse_event_type_from_nested_meta(meta)?;
+
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    let render = quote! {
+        impl#impl_generics #ident#ty_generics #where_clause {
```

Comment:
Placing `#[automatically_derived]` attribute on top of generated code is highly desired to make linters omit generated code and do not introduce another indirection of pain.